File: Data_Structures/deque.py
from array import array


class Deque:

    def __init__(self, total_size):
        self.__total_size = total_size
        self.__count = 0
        self.__main_deque = array("i", [0 for _ in range(self.__total_size)])

    def right_shift_elements(self):  # დეკის ელემენტების ერთი ბიჯით ძვრა მარჯვნივ
        for index in range(self.__total_size - 1, 0, -1):
            self.__main_deque[index] = self.__main_deque[index - 1]

    def left_shift_elements(self):  # დეკის ელემენტების ერთი ბიჯით ძვრა მარცხვნივ
        for index in range(1, self.__count):
            self.__main_deque[index - 1] = self.__main_deque[index]
        self.__main_deque[self.__count - 1] = 0

    def front(self):  # მიმართვა დეკის სათავეზე
        if self.is_empty():
            return None
        return self.__main_deque[0]

    def back(self):  # მიმართვა დეკის ბოლოზე
        if self.is_empty():
            return None
        return self.__main_deque[self.__count - 1]
    
    def push_front(self, element):  # ელემენტის ჩამატება დეკის სათავეში
        if self.is_full():
            return None
        self.right_shift_elements()
        self.__main_deque[0] = element
        self.__count += 1

    def push_back(self, element):  # ელემენტის ჩამატება დეკის ბოლოში
        if self.is_full():
            return None
        self.__main_deque[self.__count] = element
        self.__count += 1

    def pop_front(self):  # ელემენტის წაშლა დეკის სათავეში
        if self.is_empty():
            return None
        target_element = self.__main_deque[0]
        self.left_shift_elements()
        self.__count -= 1
        return target_element

    def pop_back(self):  # ელემენტის წაშლა დეკის ბოლოში
        if self.is_empty():
            return None
        target_element = self.__main_deque[self.__count - 1]
        self.__main_deque[self.__count - 1] = 0
        self.__count -= 1
        return target_element

    def size(self):  # აბრუნებს დეკში არსებულ ელემენტთა რაოდენობას
        return self.__count

    def total_size(self):  # აბრუნებს დეკში შესაძლო არსებულ ელემენტთა მაქსიმალურ რაოდენობას
        return self.__total_size

    def is_empty(self):  # გვატყობინებს არის თუ არა დეკი ცარიელი: (True/False)
        return self.__count == 0

    def is_full(self):  # გვატყობინებს არის თუ არა დეკი სავსე: (True/False)
        return self.__count == self.__total_size

    def to_list(self):  # აბრუნებს დეკს ორმაგ ბმულ სიად
        if self.is_empty():
            return []
        return [self.__main_deque[index] for index in range(self.__count)]

    def print(self):  # პრინტავს დეკში დამატებულ ელემენტებს
        for index in range(self.__count):
            print(self.__main_deque[index], end=" ")
        print()


if __name__ == "__main__":
    deque = Deque(5)
    deque.push_front(1)
    deque.push_front(2)
    deque.push_back(3)
    deque.push_back(4)
    deque.push_front(5)
    deque.pop_front()
    deque.push_front(6)
    deque.print()
    print(deque.size())
    print(deque.pop_back())
    print(deque.pop_front())
    deque.print()
    deque.push_front(10)
    deque.print()
    print("-----------")
    print(deque.size())
    print(deque.is_full())
    print(deque.is_full())
    print(deque.total_size())
    deque_to_list = deque.to_list()
    print(deque_to_list)
    print(deque.front())
    print(deque.back())

File: Data_Structures/Deque.java
import java.util.Arrays;
import java.util.NoSuchElementException;

public class Deque {

    private int totalSize;
    private int count;
    private int[] mainDeque;

    Deque() {}

    Deque(int total_size) {
        this.totalSize = total_size;
        this.count = 0;
        this.mainDeque = new int[this.totalSize];
    }

    private void rightShiftElements() { // დეკის ელემენტების ერთი ბიჯით ძვრა მარჯვნივ
        for (int index = this.totalSize - 1; index > 0; index --) {
            this.mainDeque[index] = this.mainDeque[index - 1];
        }
    }

    private void leftShiftElements() { // დეკის ელემენტების ერთი ბიჯით ძვრა მარცხვნივ
        for (int index = 1; index < this.count; index ++) {
            this.mainDeque[index - 1] = this.mainDeque[index];
        }
        this.mainDeque[this.count - 1] = 0;
    }

    private int front() throws NoSuchFieldException { // მიმართვა დეკის სათავეზე
        if (this.isEmpty()) {
            throw new NoSuchFieldException("The deque is empty");
        }
        else {
            return this.mainDeque[0];
        }
    }

    private int back() throws NoSuchFieldException { // მიმართვა დეკის ბოლოზე
        if (this.isEmpty()) {
            throw new NoSuchFieldException("The deque is empty");
        }
        else {
            return this.mainDeque[this.count - 1];
        }
    }

    private void pushFront(int element) { // ელემენტის ჩამატება დეკის სათავეში
        if (this.isFull()) {
            return;
        }
        this.rightShiftElements();
        this.mainDeque[0] = element;
        this.count ++;
    }

    private void pushBack(int element) { // ელემენტის ჩამატება დეკის ბოლოში
        if (this.isFull()) {
            return;
        }
        this.mainDeque[this.count] = element;
        this.count ++;
    }

    private int popFront() throws NoSuchFieldException { // ელემენტის წაშლა დეკის სათავეში
        if (this.isEmpty()) {
            throw new NoSuchFieldException("The deque is empty");
        }
        else {
            final int TARGET_ELEMENT = this.mainDeque[0];
            this.leftShiftElements();
            this.count --;
            return TARGET_ELEMENT;
        }
    }

    private int popBack() throws NoSuchElementException { // ელემენტის წაშლა დეკის ბოლოში
        if (this.isEmpty()) {
            throw new NoSuchElementException("The deque is empty");
        }
        else {
            final int TARGET_ELEMENT = this.mainDeque[this.count - 1];
            this.mainDeque[this.count - 1] = 0;
            this.count --;
            return TARGET_ELEMENT;
        }
    }

    private int size() { // აბრუნებს დეკში არსებულ ელემენტთა რაოდენობას
        return this.count;
    }

    private int totalSize() { // აბრუნებს დეკში შესაძლო არსებულ ელემენტთა მაქსიმალურ რაოდენობას
        return this.totalSize;
    }

    private boolean isEmpty() { // გვატყობინებს არის თუ არა დეკი ცარიელი: (True/False)
        return this.count == 0;
    }

    private boolean isFull() { //  გვატყობინებს არის თუ არა დეკი სავსე: (true/false)
        return this.count == this.totalSize;
    }

    private int[] toList() { // აბრუნებს დეკს სიად (სასარგებლო ილუზია :)))
        int[] dequeToList;
        if (this.isEmpty()) {
            return new int[0];
        }
        else {
            dequeToList = new int[this.count];
            for (int index = 0; index < this.count; index ++) {
                dequeToList[index] = this.mainDeque[index];
            }
        }
        return dequeToList;
    }

    private void print() { // პრინტავს დეკში დამატებულ ელემენტებს
        for (int index = 0; index < this.count; index ++) { // პრინტავს დეკში დამატებულ ელემენტებს
            System.out.print(this.mainDeque[index] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) throws NoSuchFieldException {
        Deque deque = new Deque(5);

        deque.pushFront(1);
        deque.pushFront(2);
        deque.pushBack(3);
        deque.pushBack(4);
        deque.pushFront(5);
        deque.popFront();
        deque.pushFront(6);
        deque.print();
        System.out.println(deque.size());
        System.out.println(deque.popBack());
        System.out.println(deque.popFront());
        deque.print();
        deque.pushFront(10);
        deque.print();
        System.out.println("---------------------");
        System.out.println(deque.size());
        System.out.println(deque.isFull());
        System.out.println(deque.totalSize());
        int [] dequeToList = deque.toList();
        System.out.println(Arrays.toString(dequeToList));
        System.out.println(deque.front());
        System.out.println(deque.back());
    }
}
